### PR TITLE
Disable server for nodereuse:false

### PR DESCRIPTION
**Pull request by rokonec**
### Context
When user requires nodereuse:false msbuild shall respect the intent of not to use any msbuild persistent processes including msbuild server.

### Changes Made
Arguments have to be preparsed, including auto response files, and msbuild server is not called when nodereuse:false and couple other cases like /help.

### Testing
Added unit test.
Running unit tests.
Local testing.

### CODE REVIEW

File: src/MSBuild/PerformanceLogEventListener.cs [hunk 1]
```diff
@@ -55,10 +55,16 @@ internal static PerformanceLogEventListener Create()
...
-                    eventListener = new PerformanceLogEventListener();
-                    eventListener.Initialize(logDirectory);
+                    if (Directory.CreateDirectory(logDirectory).Exists)
+                    {
+                        eventListener = new PerformanceLogEventListener();
+                        eventListener.Initialize(logDirectory);
+                    }
+
+                    // Delete env var so new starting processes will not duplicate Event listening
+                    Environment.SetEnvironmentVariable(PerfLogDirEnvVar, null);
```
Comment (Forgind) [inline src/MSBuild/PerformanceLogEventListener.cs:67]: Do you have to reset this at the end of the build?

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -266,6 +274,57 @@ string[] args
...
+                    canRunServer = false;
+                    if (KnownTelemetry.BuildTelemetry != null)
+                    {
+                        KnownTelemetry.BuildTelemetry.ServerFallbackReason = "Arguments";
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                CommunicationsUtilities.Trace("Unexpected exception during command line parsing. Can not determine if it is allowed to use Server. Fallback to old. Exception: {0}", ex);
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:317]:
```suggestion
                CommunicationsUtilities.Trace("Unexpected exception during command line parsing. Can not determine if it is allowed to use Server. Fall back to old behavior. Exception: {0}", ex);
```

File: src/MSBuild/XMake.cs [hunk 3]
```diff
@@ -266,6 +274,57 @@ string[] args
...
+            string commandLine
+#else
+            string[] commandLine
+#endif
+            )
+        {
+            bool canRunServer = true;
+            try
+            {
+                GatherAllSwitches(commandLine, out var switchesFromAutoResponseFile, out var switchesNotFromAutoResponseFile, out string fullCommandLine);
```
Comment (Forgind) [inline src/MSBuild/XMake.cs:295]: I think this shouldn't have much perf impact because we only would do this once at the start of the build, but I'm wondering about small builds where even a little extra perf can matter. Can we reuse what we parsed here?

**Review by Forgind — APPROVED** [review]
LGTM!

**Review by AR-May — APPROVED** [review]

File: src/MSBuild/PerformanceLogEventListener.cs (hunk 1 above)
Comment (rokonec) [reply src/MSBuild/PerformanceLogEventListener.cs]: On second, morning, thought I have reverted that change. I though that it was cross process event listener but it is actually in-process event listener which needs to run for every process involved.
TLDR: It was good as it was.

File: src/MSBuild/XMake.cs (hunk 3 above)
Comment (rokonec) [reply src/MSBuild/XMake.cs:295]: I measured it and it is under 1ms. 
But that being said, whole argument parsing would benefit refactoring - for clarity. I would strongly suggest to split argument parsing from interpreting for example. 
But that shall be addressed in different PR.

**Review by Forgind — APPROVED** [review]
